scripts/linters/pylint_extensions.py

```diff
@@ -1654,99 +1654,88 @@ class SingleLineCommentChecker(checkers.BaseChecker):
+                line = line.strip()
 
-            # Comments may include a lowercase character at beginning
-            # or may not use a punctuation at end if it contains a
-            # excluded phrase e.g. "# coding: utf-8".
-            if word_is_present_in_excluded_phrases:
-                continue
+                if not line.startswith('#'):
+                    continue
 
-            if not next_line.startswith(b'#'):
-                # Comments must end with the proper punctuation.
-                last_char_is_invalid = line[-1] not in (
-                    ALLOWED_TERMINATING_PUNCTUATIONS)
-                if last_char_is_invalid:
+                # Comments must start with a space.
+                if (re.search(br'^#[^\s].*$', line) and not
+                        line.startswith(b'#!')):
+                    space_at_beginning_of_comment = False
                     self.add_message(
-                        'invalid-punctuation-used', line=line_num + 1)
+                        'no-space-at-beginning', line=line_num)
 
-            # Comments must start with a space.
-            if re.search(br'^#[^\s].*$', line) and not line.startswith(b'#!'):
-                space_at_beginning_of_comment = False
-                self.add_message(
-                    'no-space-at-beginning', line=line_num + 1)
-
-            # Comments may include a lowercase character at the beginning
-            # only if they start with version info or a data type or.
-            # a variable name e.g. "# next_line is of string type."
-            # or "# v2 version does not have ExplorationStatsModel."
-            # or "# int. The file size, in bytes.".
-            if len(line) > 1 and space_at_beginning_of_comment:
-
-                # Check if variable name is used.
-                underscore_is_present = '_' in line.split()[1]
-                if underscore_is_present:
-                    continue
+                if line_num - previous_line_num > 1:
+                    split_line = line.split()
+                    split_prev_line = previous_line.split()
+
+                    # Check if comment contains any excluded phrase.
+                    excluded_phrase_is_present_at_beginning = any(
+                        word in split_line[1] for word in
+                        EXCLUDED_PHRASES)
+
+                    if len(previous_line) > 1:
+                        excluded_phrase_is_present_at_end = any(
+                            word in split_prev_line[-1] for word in
+                            EXCLUDED_PHRASES)
+                        excluded_phrase_at_beginning_of_prev_line = any(
+                            word in split_prev_line[1] for word in
+                            EXCLUDED_PHRASES)
+
+                        # Comments must end with the proper punctuation.
+                        last_char_is_invalid = previous_line[-1] not in (
+                            ALLOWED_TERMINATING_PUNCTUATIONS)
```

Comment:
Why are we checking punctuations in previous line? I'm finding it complicated to understand the structure of the code! :( Can you please explain the basic structure?

Reply:
We are only checking the first and the last line of a comment block. So we get to the starting of a new comment block, we have to check if the last line ended with proper punctuation.

Example:
```
# Comment.

# Comment.
```

Reply:
If only the first and last line matters, then do you think this structure will be more readable and clear:

```
Create a list of (one-line and multiline) comments in the file using the token: [[online-comment-token], [1st-linecomment-token, 2nd-line-comment-token], ...]

For each comment:
   Do all the required checks.
```

Do you think this will be more clear and readable? i.e, we won't have to create lots of variables to track the previous line and the previous line comments? what do you think?

Reply:
I think it will make it more complicated, as we have to separate single-line comments and multiline comments then and then run checks on those comments.

Reply:
We have lots of variables to update and compute here (like prev_line, prev_line_no etc.) and it's throughout this function. 

> as we have to separate single-line comments and multiline comments then and then run checks on those comments.

We will just do it once on top and we don't have to trace previous line related things. Once the comments token variable is ready we just have to focus on each comment. Let me know what you think?

Reply:
I've changed the method.